src/tic-tac-toe.ts

```diff
@@ -40,35 +43,73 @@ class Game {
     return square === this.currentPlayer;
   };
 
-  private checkRowsForWin(): boolean {
-    return this.board.some(row => row.every(this.isCurrentPlayer));
+  private checkRowsForWin(): WinningVector | undefined {
+    for (let rowIndex = 0; rowIndex < this.board.length; rowIndex++) {
+      const row = this.board[rowIndex];
+      const vector: WinningVector = [];
+      for (let colIndex = 0; colIndex < this.board.length; colIndex++) {
+        const square = row[colIndex];
+        if (this.isCurrentPlayer(square)) {
+          vector.push({ row: rowIndex, col: colIndex });
+        }
+      }
+      if (vector.length === this.boardSize) {
+        return vector;
+      }
+    }
+    return;
   }
 
-  private checkColumnsForWin(): boolean {
+  private checkColumnsForWin(): WinningVector | undefined {
     const transposedBoard = this.eachIndex().map(i =>
       this.board.map(row => row[i])
     );
 
-    return transposedBoard.some(row => row.every(this.isCurrentPlayer));
+    for (let colIndex = 0; colIndex < transposedBoard.length; colIndex++) {
+      const column = transposedBoard[colIndex];
+      const vector: WinningVector = [];
+
+      for (let rowIndex = 0; rowIndex < transposedBoard.length; rowIndex++) {
+        const square = column[rowIndex];
+        if (this.isCurrentPlayer(square)) {
+          vector.push({ row: rowIndex, col: colIndex });
+        }
+      }
+      if (vector.length === this.boardSize) {
+        return vector;
+      }
+    }
+    return;
   }
 
-  private checkDiagonalsForWin(): boolean {
-    const diagonals = [
-      this.eachIndex().map(i => this.board[i][i]),
-      this.eachIndex().map(i => this.board[i][this.boardSize - 1 - i]),
-    ];
+  private checkDiagonalsForWin(): WinningVector | undefined {
+    const winningVector: WinningVector = [];
 
-    return diagonals.some(values => values.every(this.isCurrentPlayer));
+    for (let i = 0; i < this.boardSize; i++) {
+      // check values from top left to bottom right (ie for 3x3 grid => [0, 0], [1, 1], [2, 2])
+      if (this.isCurrentPlayer(this.board[i][i])) {
+        winningVector.push({ row: i, col: i });
+        // check values from top right to bottom left (ie for 3x3 grid => [0, 2], [1, 1], [2, 0])
+      } else if (this.isCurrentPlayer(this.board[i][this.boardSize - 1 - i])) {
+        winningVector.push({ row: i, col: this.boardSize - 1 - i });
+      }
+      if (winningVector.length === this.boardSize) {
+        return winningVector;
+      }
+    }
+    return;
   }
 
   private checkGameForWin(): void {
-    const winCondition = [
+    const winVectors = [
       this.checkRowsForWin(),
       this.checkColumnsForWin(),
-      this.checkDiagonalsForWin()
+      this.checkDiagonalsForWin(),
     ];
 
-    if (winCondition.some(winnerCheck => winnerCheck)) {
+    const maybeWinningVector = winVectors.find(vector => !!vector);
+    if (maybeWinningVector) {
+      this.winningVector = maybeWinningVector;
```

Comment:
I disagree with pushing the null checks downstream. We talked about this and decided that a "win" is a winningVector with a nonzero length; changing downstream calls from `winningVector.length` to `winningVector && winningVector.length` does not seem useful to me. 

```jsx
// in checkGameForWin:

this.winningVector = winVectors.find(vector => vector && vector.length) || [];
```

Reply:
I updated it to the below, but get a type error I don't understand.

```
private checkGameForWin(): void {
    const winVectors = [
      this.checkRowsForWin(),
      this.checkColumnsForWin(),
      this.checkDiagonalsForWin(),
    ];

    this.winningVector =
      winVectors.find(vector => vector && vector.length) || [];

    if (this.winningVector.length >= this.boardSize) {
      this.winner = this.currentPlayer;
    }
  }
}
```


```
Type 'undefined' is not assignable to type 'boolean'.

111    winVectors.find(vector => vector && vector.length) || [];

node_modules/typescript/lib/lib.es2015.core.d.ts:32:21
32    find(predicate: (value: T, index: number, obj: T[]) => boolean, thisArg?: any): T | undefined;
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
The expected type comes from the return type of this signature.
```

Changing the `this.winningVector` assignment line to the below works though:
```
this.winningVector = winVectors.find(vector => !!vector) || [];
```